ebos/eclwriter.hh

```diff
@@ -281,10 +302,30 @@ public:
     void writeOutput(bool isSubStep)
     {
         const int reportStepNum = simulator_.episodeIndex() + 1;
-
         this->prepareLocalCellData(isSubStep, reportStepNum);
         this->eclOutputModule_->outputErrorLog(simulator_.gridView().comm());
+#ifdef HAVE_DAMARIS
+        if (EWOMS_GET_PARAM(TypeTag, bool, EnableDamarisOutput)) {
+            // N.B. damarisUpdate_ should be set to true if at any time the model geometry changes
+            if (this->damarisUpdate_) {
+                const auto& gridView = simulator_.gridView();
+                const int numElements = gridView.size(/*codim=*/0); // This it is the local ranks model size.
+                Opm::DamarisOutput::setupDamarisWritingPars(simulator_.vanguard().grid().comm(), numElements);
+                // By default we assume static grid
+                this->damarisUpdate_ = false;
+            }
 
+            if (!isSubStep) {
+                data::Solution localCellData = {};
+                this->eclOutputModule_->assignToSolution(localCellData);
+                // Output the PRESSURE field
+                if (this->eclOutputModule_->getPRESSURE_ptr() != nullptr) {
+                    damaris_write("PRESSURE", (void*)this->eclOutputModule_->getPRESSURE_ptr());
+                    damaris_end_iteration();
+                }
```

Comment:
I didn't pick up on this before, but this sequence of statements risks reading from moved-from memory.  The reason is that [`assignToSolution()`](https://github.com/OPM/opm-simulators/blob/2064e8725f11a5ae00e0fd96a686809fd6eab983/ebos/eclgenericoutputblackoilmodule.cc#L645) currently [transfers](https://github.com/OPM/opm-simulators/blob/2064e8725f11a5ae00e0fd96a686809fd6eab983/ebos/eclgenericoutputblackoilmodule.cc#L655) ownership of its solution arrays into the `data::Solution` object, including that of the array that holds [`PRESSURE`](https://github.com/OPM/opm-simulators/blob/2064e8725f11a5ae00e0fd96a686809fd6eab983/ebos/eclgenericoutputblackoilmodule.cc#L683) values.  There is also no guarantee that `vector::data()` is `nullptr` after it has been moved from.

Furthermore, by calling `assignToSolution()` and just discarding the `localCellData`, you remove all the data arrays that would otherwise go to the ECLIPSE-compatible restart file.  Is that really intentional?  My suggestion here would be to remove the call to `assignToSolution()` here and just use the `PRESSURE_ptr()` functions exclusively.

Reply:
Thanks for the explanation. I agree that the assignToSolution() is incorrect here. The PRESSURE_ptr() is a temporary API for testing, I think long term we should get the output arrays from the data::Solution object. Then it may be correct to reinstate it, or rather to make sure we only do assignToSolution() once by unifying with the parts further down in the code.

I will test and push a PR to remove the assignToSolution() call now.